### PR TITLE
Implement pools

### PR DESCRIPTION
**Pull request by notjames**
Added support for pools. Still really new to Go so constructive assistance appreciated.

### CODE REVIEW

File: machine.go [hunk 1]
```diff
@@ -5,12 +5,11 @@ package gomaasapi
 
 import (
 	"fmt"
-	"net/http"
-	"net/url"
-
 	"github.com/juju/errors"
 	"github.com/juju/schema"
 	"github.com/juju/version"
+	"net/http"
```
Comment (wallyworld) [inline machine.go]: Please move thee imports up to the first block containing those from the std library

File: machine.go [hunk 2]
```diff
@@ -575,6 +594,7 @@ func machine_2_0(source map[string]interface{}) (*machine, error) {
 		bootInterface:        bootInterface,
 		interfaceSet:         interfaceSet,
 		zone:                 zone,
+		pool:      			  pool,
```
Comment (wallyworld) [inline machine.go]: please run go fmt to correct spacing etc

File: pool_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        "description": "default description",
+        "resource_uri": "/MAAS/api/2.0/pools/default/",
+        "name": "default"
+    }, {
+        "description": "special description",
+        "resource_uri": "/MAAS/api/2.0/pools/special/",
+        "name": "special"
+    }
+]
+`
```
Comment (wallyworld) [inline pool_test.go]: please add a new line at the end of each file

File: pool.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
+// Copyright 2016 Canonical Ltd.
```
Comment (wallyworld) [inline pool.go]: 2019

File: pool_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2016 Canonical Ltd.
```
Comment (wallyworld) [inline pool_test.go]: 2019

File: controller.go [hunk 6]
```diff
@@ -222,6 +222,26 @@ func (c *controller) Zones() ([]Zone, error) {
 	return result, nil
 }
 
+// Pools implements Controller.
+func (c *controller) Pools() ([]Pool, error) {
```
Comment (wallyworld) [inline controller.go:226]: Please add a test to controller_test.go similar to the zone tests that are already there

File: pool.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	return result, nil
+}
+
+type poolDeserializationFunc func(map[string]interface{}) (*pool, error)
+
+var poolDeserializationFuncs = map[version.Number]poolDeserializationFunc{
+	twoDotOh: Pool2dot0,
+}
+
+func Pool2dot0(source map[string]interface{}) (*pool, error) {
```
Comment (wallyworld) [inline pool.go]: this doesn't need to be exported, so rename to pool_2_0
(to be consistent with zone)

**Review by wallyworld — COMMENTED** [review]
This looks great, thank you.
Just a few small things to tweak. Main issue is to add controller tests.

File: machine.go (hunk 1 above)
Comment (notjames) [reply machine.go]: done

File: pool_test.go (hunk 3 above)
Comment (notjames) [reply pool_test.go]: done

**Comment by notjames** [comment]
I will get the rest of these done either tonight or tomorrow. I also found a bug I need to fix where a nil Pool returns nil instead of an empty map. I'll fix that too. Thank you for reviewing!

**Comment by notjames** [comment]
had a production impacting issue occur today so likely won't get back to this until tomorrow. It's at the top of my non-prod list of things to do though.

File: machine.go (hunk 1 above)
Comment (wallyworld) [reply machine.go]: The PR still shows these imports grouped into the wrong block?

**Review by wallyworld — COMMENTED** [review]
Just noticed some new commits. Thank you.
We need a TestPools() similar to TestZones() in controller_test.go
Also, some imports need moving.

**Comment by notjames** [comment]
I understand. Those commits were not meant to be checked yet. Sorry. My IDE is responsible for the imports. I'll fix all of that up. I am hoping to have this done today provided I'm able to fix the test bug I exposed. I'm also going to rebase/squash all the commits. When that is done then you'll know I'm finished and ready for full review.

**Comment by notjames** [comment]
This PR is ready for final review and merge.

File: interface_test.go [hunk 8]
```diff
@@ -471,7 +471,7 @@ const (
     "parents": ["bond0"],
     "id": 40,
     "type": "physical",
-    "resource_uri": "/MAAS/api/2.0/nodes/4y3ha6/interfaces/40/",
+    "resource_uri": "/maas/api/2.0/nodes/4y3ha6/interfaces/40/",
```
Comment (wallyworld) [inline interface_test.go]: Curious about this change.
It's "/MAAS/..." elsewhere

**Review by wallyworld — APPROVED** [review]
Thanks! Just one small correction. Then you can merge by typing
$$merge$$
as a comment

File: interface_test.go (hunk 8 above)
Comment (notjames) [reply interface_test.go]: Looks like just an IDE auto-edit. Fixing.

**Comment by notjames** [comment]
$$merge$$

**Comment by notjames** [comment]
merge

**Comment by notjames** [comment]
Hmm, the merge macro seems to not work...?

**Comment by mitechie** [comment]
I don't think it'll work for you. The bot wants someone from the team to do it. 

$$merge$$

**Comment by notjames** [comment]
Makes sense! Thank you again @mitechie (Rick)

**Comment by stgraber** [comment]
So this broke LXD's MAAS support as any attempt to create a device without the `pool` property set now gives us: 

```
DBUG[07-27|16:44:52] Failure for task operation: a5c57abb-4458-4f16-9027-c1eda178cbb4: Create container: Create LXC container: device 0: device 2.0 schema check failed: pool: expected map, got nothing 
```

**Comment by stgraber** [comment]
One of the failing code paths is when CreateDevice is called on a machine, this in turns calls CreateDevice on the controller, neither of those pass a pool (not needed as there is a parent), the response coming from MAAS similarly doesn't include a pool (it's nil) and it being nil causes a validation failure.

**Comment by stgraber** [comment]
This seems to indicate that CreateDevice on a machine isn't tested or was never tested against an actual MAAS server as I can't see any way this would ever work.

**Comment by notjames** [comment]
Good catch! I don't have time today to look into it, but I'll try and dig in tomorrow.

**Comment by stgraber** [comment]
@notjames https://github.com/juju/gomaasapi/pull/80 seemed to work for me, but that's only tested for the subset of features that LXD uses

**Comment by notjames** [comment]
@stgraber I will take a look and make sure everything works. We have a few maas implementations so if I can find time, I'll do some more testing.

**Pull request merged**